
# <p style="text-align: center;">EE 461P: Data Science Principles</p>
# <p style="text-align: center;">Assignment 1</p>
## <p style="text-align: center;">Total points: 90</p>
## <p style="text-align: center;">Due: Tues, September 18, submitted via Canvas by 11:59 pm</p>

Your homework should be written in a **Jupyter notebook**. You may work in groups of two if you wish. Only one student per team needs to submit the assignment on Canvas.  But be sure to include name and UTID for both students.  Homework groups will be created and managed through Canvas, so please do not arbitrarily change your homework group. If you do change, let the TAs know.

Also, please make sure your code runs and the graphics (and anything else) are displayed in your notebook before submitting. (%matplotlib inline)

# Question 1: Applications of machine learning (10 pts)

Read the [article](http://www.datasciencecentral.com/profiles/blogs/20-data-science-systems-used-by-amazon-to-operate-its-business) "21 data science systems used by Amazon to operate its business" and pick any two of the data science systems used by Amazon according to this blog.

(5 pts each) For each of these two systems you have chosen:

What kind of machine learning problem is involved (e.g. classification, regression, clustering, outlier detection,...)? Speculate on what kind of data may be needed and how the results can be useful to the company.


## Answer



# Question 2.a: Maximum likelihood estimate (10 pts)

Suppose a manager at an internet sales company wants to estimate how fast his salesperson is generating successful leads. Instead of recording the time for each lead, the time taken to generate the next 3 leads are recorded, i.e., there is one recording (denoting the elapsed time) for every 3 consecutive leads. For a specific salesperson, the time intervals recorded are {1,3,1.5,4,2,7,1.2,2,4,3.1} hours. 

A statistician suggests that if these time intervals are assumed to arise by i.i.d. sampling from the following distribution:
$$ p(t) = \frac{1}{C \times \theta^{3}}t^{2}exp^{-\frac{t}{\theta}},$$
(where C is a normalizing constant). Therefore, if $\theta$ can be estimated, then he can provide detailed information
about the lead generation process, including average rates, variances etc.

Find the Maximum Likelihood estimate for $\theta$ based on the recorded observations.


## Answer

Assume iid observations of lead times and the distribution recommended by the statistician. Let $i$ index the $i^{th}$ observation $t_i$; the likelihood of $n$ observations is given by:

\begin{align*}
\begin{split}
L(\theta) = \prod_{i = 1}^n \bigg( \frac{1}{C \theta^3} t_i^{2}\exp^{-\frac{t_i}{\theta}}\bigg)\\
\text{Taking log on both sides, we have,}\\
\log{L(\theta)} &= \sum_{i = 1}^n \bigg( \frac{1}{C \theta^3} t_i^{2}\exp^{-\frac{t_i}{\theta}}\bigg) \\
&= \sum_{i = 1}^n \bigg( 2t_i - \frac{t_i}{\theta} -C - 3\theta \bigg)\\
\text{Taking derivative w.r.t $\theta$ gives}\\
\frac{\partial \log{L(\theta)}}{\partial \theta}&= \frac{\sum_{i = 1}^n t_i }{\theta^2} - \frac{3n}{\theta} \\
\end{split}
\end{align*}
Notice that first and third terms are not functions of $\theta$ and the last term is subtracted $n$ times. Solving for $\theta$ by setting the derivative to 0, we have, 
\begin{equation*}
\theta = \frac{\sum_{i = 1}^n t_i}{3n}
\end{equation*}


Substituting $n = 10$ and $t_i$ to the observations provided, we have, $\theta = 0.96$.

Note the distribution used to model lead times is called a gamma distribution. The distribution is a generalization to the exponential distribution (usually used to model inter-arrival times between events modeled as a Poisson process. Sum of 3 inter-arrival times will be Gamma with k=3, which is what I substituted in the (two parameter) Gamma distribution. You can read the details here: https://en.wikipedia.org/wiki/Gamma_distribution.

# Question 2.b: Bivariate Normal Distribution (10 pts)

Suppose $X$ and $Y$ are two random variables whose joint distribution is Normal (Gaussian), centered at (0,0) and with correlation $\rho$. (See "Bivariate Case" in the Wikipedia entry for "Multivariate Normal Distribution" for the equation, or use just use the vector form given in the class notes, with $\sigma_{12} = \sigma_{21} = \rho \sigma_{x} \sigma_{y} $). Consider 2 cases:
1. $\sigma_x^2 = 16; \sigma_y^2 = 9; \rho = 0$
2. $\sigma_x^2 = 16; \sigma_y^2 = 9; \rho = 0.5$

a. (4 pts) Obtain contour plots for each of the two distributions using  Python (https://seaborn.pydata.org/generated/seaborn.kdeplot.html). 

b. (3 pts) View 3-D plots for the two distributions from at least two different viewing perspectives each (http://matplotlib.org/examples/mplot3d/rotate_axes3d_demo.html).

c.  (3 pts) Consider the bivariate Normal Distribution given in part (ii). We can rotate this distribution by using the rotation matrix
\begin{equation} \left[ \begin{array}{cc} \cos(\theta) & -\sin(\theta) \\ \sin(\theta) & \cos(\theta) \end{array} \right] \end{equation}
Check Wikipedia for more info on rotation matrices. This is the same as creating two new random variables:
\begin{aligned}
U = X \cos(\theta) - Y \sin(\theta) \\
V = X \sin(\theta) + Y \cos(\theta)
\end{aligned}
For a rotation of -45 degrees, plot the rotated distribution and fit a bivariate gaussian to this rotated distribution.

## Answer


# Question 3: Multiple Linear Regression in Python (25 pts)

Use the following code to import data that can be used for predicted the popularity of online news articles.



In [23]:
from sklearn import datasets, linear_model
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd

# load database
data = pd.read_csv('data_q3.csv')

# show the first 5 rows of DB
pd.DataFrame.head(data)

n_tokens_title  log_n_tokens_content  n_unique_tokens  \
0              12              5.662960         0.589474   
1              10              6.025866         0.606796   
2              11              4.897840         0.702290   
3              11              5.638355         0.610714   
4              12              6.212606         0.513026   

   n_non_stop_unique_tokens  sqrt_num_hrefs  sqrt_num_self_hrefs  \
0                  0.726190        2.236068             1.414214   
1                  0.777344        2.828427             0.000000   
2                  0.785714        2.000000             1.414214   
3                  0.766467        2.236068             2.000000   
4                  0.662252        3.741657             1.000000   

   sqrt_num_imgs  sqrt_num_videos  average_token_length  num_keywords  \
0            0.0              0.0              4.381944             6   
1            1.0              0.0              4.980676             6   
2            3.0              0.0              4.477612            10   
3            1.0              0.0              4.608541             4   
4            1.0              0.0              5.084168            10   

      ...      min_positive_polarity  max_positive_polarity  \
0     ...                   0.166667                    1.0   
1     ...                   0.100000                    0.5   
2     ...                   0.136364                    0.8   
3     ...                   0.136364                    0.5   
4     ...                   0.100000                    0.5   

   avg_negative_polarity  min_negative_polarity  max_negative_polarity  \
0              -0.366667                  -0.80                  -0.10   
1              -0.176128                  -0.70                  -0.05   
2              -0.150000                  -0.15                  -0.15   
3              -0.278472                  -0.50                  -0.05   
4              -0.235000                  -0.50                  -0.10   

   title_subjectivity  title_sentiment_polarity  abs_title_subjectivity  \
0                 0.0                       0.0                     0.5   
1                 0.3                       0.2                     0.2   
2                 0.0                       0.0                     0.5   
3                 0.0                       0.0                     0.5   
4                 0.3                       0.0                     0.2   

   abs_title_sentiment_polarity  log_shares  
0                           0.0    6.626718  
1                           0.2    6.148468  
2                           0.0    8.433812  
3                           0.0    7.170120  
4                           0.0    7.090077  

[5 rows x 40 columns]

In [24]:
# build inputs (features) and output
X = data.iloc[:,:-1]
y = data.iloc[:,-1]

# standardization (rendering data zero mean and standard deviation 1)
X_norm = (X - X.mean()) / X.std()

# split DB into training-set and test-set
msk = np.random.rand(len(data)) < 0.8
X_train = X_norm[msk]
y_train = y[msk]

X_test = X_norm[~msk]
y_test = y[~msk]

(a)  (5 pts) Using ordinary least squares, fit a multiple linear regression (MLR) on all the feature variables using the entire dataset. Report the regression coefficient of each input feature and evaluate the model using mean squared error (MSE).  Example of ordinary least squares in Python is shown in Section 1.1.1 of http://scikit-learn.org/stable/modules/linear_model.html.

(b)  (5 pts) Fit an MLR using the training set.  Evaluate the trained model using the training set and the test set, respectively. Compare the two MSE values.

(c)  (5 pts) Do you think your MLR model is reasonable for this problem? You may look at the distribution of residuals to provide an informed answer.

(d) (5 pts) Identify and list the the three most siginificant features. You might find this link to be helpful: http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html.

(e) (5 pts) Build a new MLR model using only the three most significant features. Again, use 80% of the data for training and the remaining 20% for testing. Then, Compare the test set MSE values from (b) and (e). Please explain what you observe.

## Answer



# Question 4: Ridge and Lasso Regression (25 points)
Use the following codes to import the already-standardized data.

In [1]:
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
import numpy as np
import pandas as pd

# load database
filename = "data_q4.csv"
data = pd.read_csv(filename)

X = data.values[:,0:8]
y = data.values[:,-1]

The dataset consists of eight features and one target variable.

After importing the datasets, use the following code to add additional second-order polynomial and interaction variables to the features matrix.  You should have 44 variables and one target variable. Note that this code adds all combinations of the features with degree less than or equal to two; in practice one may introduce only a few based on domain knowledge or experimentation.

In [18]:
from sklearn.preprocessing import PolynomialFeatures
X = PolynomialFeatures(2, include_bias=False).fit_transform(X)

In this question, you will explore the application of Lasso and Ridge regression using the sklearn package in Python. The following code will split the data into training and test set using [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) with **random state 20** and **test_size = 0.33**.  Note: lambda is called alpha in sklearn.

In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state=20)

a) (8pts) Use sklearn.linear_model.Lasso and sklearn.linear_model.Ridge classes to do a [5-fold cross validation](http://scikit-learn.org/stable/auto_examples/exercises/plot_cv_diabetes.html#example-exercises-plot-cv-diabetes-py) using sklearn's [KFold](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html#sklearn.model_selection.KFold). For the sweep of the regularization parameter, we will look at a grid of values ranging from $\lambda = 10^{10}$ to $\lambda = 10^{-5}$. In Python, you can consider this range of values as follows:

      import numpy as np

      alphas =  10**np.linspace(10,-5,100)

  Report the best chosen $\lambda$ based on cross validation. The cross validation should happen on your training data using  average MSE as the scoring metric. 

b) (8pts) Run ridge and lasso for all of the alphas specified above (on training data), and plot the coefficients learned for each of them - there should be one plot each for lasso and ridge, so a total of two plots; the plots for different features for a method should be on the same plot (e.g. Fig 6.6 of JW). What do you qualitatively observe when value of the regularization parameter is changed? 

c) (5pts) Run least squares regression, ridge, and lasso on the training data. For ridge and lasso, use only the best regularization parameter. Report the prediction error (MSE) on the test data for each. 

d) (4pts) Run lasso again with cross validation using [sklearn.linear_model.LassoCV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LassoCV.html). Set the cross validation parameters as follows:

    LassoCV(alphas=None, cv=10, max_iter=10000)

Report the best $\lambda$ based on cross validation. Run lasso on the training data using the best $\lambda$ and report the coefficeints for 65 variables. What do you observe from these coefficients? 

## Answer



# Question 5 (10 pts)
Please solve problem \#3.3 on page 174 in Bishop (Chapter 3). The problem has been uploaded to Canvas under 'Files': **CH3 problems from Bishop PRML.pdf**.

## Answer